Automattic/isolated-block-editor
# Collab: Add support for same-block concurrent editing

Closes #39
Closes #60

## Background

The existing implementation worked well as long as multiple users were editing _different_ blocks at the same time, but it did not work when they edited the _same_ block at the same time. There were two problems:

1. Your cursor could be shifted by an incoming peer edit. For example, if you were typing at cursor index 3 (`abc|`) and somebody prepended a character (`xabc`), cursor index 3 would now be `xab|c`. This would disrupt your typing.
2. Simultaneous edits could not be merged naturally. For example, let's say two users started with the letter `b` in their editor. The first user prepended an `a` and the second user appended a `c` at the same time. When these changes converge, we'd want it to merge into `abc`. In reality, this would only converge into either `ab` or `bc`, and one of the users would perceive it as their changes "backtracking".

Both problems stemmed from the fact that Gutenberg only exposes the "result" of every edit, and not the "operation" that led to that result. So if our initial state is `b`, the operation and result would be like:

- Operation: "Insert `a` at index `0`"
- Result: `ab`

In order to solve the cursor shifting and text merging problems, we need to manipulate our Yjs data with index-based insert/delete operations, instead of just bulk overwriting the result for every edit. Because Gutenberg does not expose these operations, we need to "reverse engineer" each operation by taking the diff of the before and after states.

## Is this a good idea?

The work in this PR is exploratory in nature. It uncovers where the gaps are for full collab support in Gutenberg, more than it aims to be a permanent solution.

### Performance

Although there are ways we can optimize it a little more, the code does a lot of work that could be avoided if we could get operation data directly from Gutenberg. It's certainly not ideal, but I'm interested in how usable this is in real life.

### Limitations

There are some minor limitations, flagged in the code with the `// TODO: Unsolved problem` comment. Some are due to limitations in Yjs, and some are due to unexposed data in Gutenberg.

## Testing instructions

1. `yarn storybook`
2. Open two browser tabs at http://localhost:6006/?path=/story/collaboration--default
3. In one tab, type a sentence and keep your cursor near the end of the line.
4. In the other tab, prepend some text to the same sentence. The peer cursor should not shift erratically.
5. To test truly simultaneous edits, you can't do it in Storybook and you need two devices. Automatticians can try `arc patch D73765` with colleagues.
